### PR TITLE
Move the block movers to the block toolbar

### LINKED ISSUE #16617: Move buttons are hard to use within a nested block
**Describe the bug**
Using a meda+text block, its hard to move the paragraphs around without actually focusing on them — the move buttons appear when hovered, but you can't actually interact with them. Here's a GIF that hopefully helps you understand the problem:

confusing buttons

Alternatively, sometimes the opposite is true — the move buttons get in the way of resizing the image. Here's another GIF:

confusing buttons 2

The behavior seems to change depending on what block you currently have focused, but I can't seem to understand the logic behind the change.

### LINKED ISSUE #18069: Block movers are funky when block is align-right or align-left
**Describe the bug**
The block movers get a bit funky when the block is set for align-right or align-left. They appear with increased padding?

**To reproduce**
1. Create an Image block.
2. Set block to align-right or align-left.
3. Observe the block movers and notice the increase in size.

**Expected behavior**
I expect the block movers to never change size when the block is aligned differently.

**Screenshots**
_having trouble uploading image, so here's a link_

https://cldup.com/2omdT_rdWq.png

**Desktop (please complete the following information):**
 - OSX
 - Firefox 69.0.2
 - Gutenberg 6.7

### LINKED ISSUE #18195: Column Block: Movers for Full and Wide Images inside columns are not visible
**Describe the bug**
When you add a Full or Wide image inside a Column Block, the movers are not visible:

Screen Shot 2019-10-30 at 1 24 02 PM

**To reproduce**
Steps to reproduce the behaviour:
1. Create a new Column Block
2. Add an Image block inside a column
3. Set the image to Wide or Full
4. The Mover will not be visible

**Expected behaviour**
I should be able to see and use the Mover

**Screenshots**
With a "Full" width column block and a Full width image:
Screen Shot 2019-10-30 at 1 25 58 PM

With a "Full" width column block and a Wide image:
Screen Shot 2019-10-30 at 1 28 20 PM

**Desktop (please complete the following information):**
 - OS: macOS Mojave
 - Browser: Chrome
 - Version 78.0.3904.70 (Official Build) (64-bit)

**Additional context**
- Using latest `master`

### LINKED ISSUE #18223: Block Toolbars: Reposition block toolbars to ensure they render within the viewport
image

Depending on the viewport width, the placement of the block, and the length of the toolbar — its not uncommon to see the toolbar renders outside the visible space on the screen. Right now this adds a horizontal scrollbar to the entire document.

Instead, I'd expect the toolbar to move itself to ensure it displays within the viewport. Something like this:

image

### PR DESCRIPTION
**Pull request by youknowriad**
closes #18195 
closes #18069 
closes #18223
closes #16617 

Related #18667 #18632 

This PR moves the block movers to the block toolbar consistently.

### CODE REVIEW

**Comment by mtias** [comment]
Can we also combine with https://github.com/WordPress/gutenberg/pull/17352 for the horizontal variant to see how it'd look code-wise?

**Comment by mtias** [comment]
It's nice to see the reduction in code, this should help us maintain and have more consistent interactions :)

**Comment by ZebulanStanphill** [comment]
There seems to be this weird bug that happens now. Try creating a columns block and dragging a block in one of the columns to the other column. Suddenly the selected block gets stuck in this weird state where its toolbar is gone and its content is invisible. This does not happen in `master`.

image

Aside from that issue, everything else seems to be working fine. I think the drag handle size could be reduced, especially if the move buttons also doubled as drag handles.

I think it would be helpful if the toolbar could extend to the left of the block border in tight situations like this one:

image

However, that's probably out of the scope of this PR.

**Comment by youknowriad** [comment]
@ZebulanStanphill 

> I think it would be helpful if the toolbar could extend to the left of the block border in tight situations like this one:

Yes, hard to get right though in a generic way.

--
I pushed a fix for the reported bug.

**Comment by ZebulanStanphill** [comment]
I've confirmed that the dragging bug is fixed.

Something else I just noticed is that mobile now uses the same sibling inserter as the desktop. Also, the toolbar now overlaps the block below on mobile.

`master`:
image

This PR:
image

I think something needs to be done about the sibling inserter. I also prefer the no-overlap behavior of the `master` mobile toolbar.

**Comment by youknowriad** [comment]
Thanks for the feedback @ZebulanStanphill 

I actually didn't touch the sibling inserter at all, I believe you can still trigger it in master as well. I do think the in-between inserter is not that useful in mobile especially because the inserter is always accessible easily (top of the screen) and has the same behavior but I'm not going to touch these in this particular PR.

I'll have to check where the extra padding was added for the toolbar before and restore it.

**Comment by ZebulanStanphill** [comment]
Ah, I had forgotten about the add button in the top toolbar. Good point.

**Comment by ZebulanStanphill** [comment]
With the recent changes to the mobile toolbar, everything seems to be working fine in this branch now.

**Comment by ItsJonQ** [comment]
@youknowriad Hallooo!

I tested your update :). I've also been actively testing interactions via [prototypes](https://codesandbox.io/s/github/ItsJonQ/g2-prototypes/tree/master/) from @mtias 's post on [Advancing the Block Interface](https://github.com/WordPress/gutenberg/issues/18667).

I'm not 100% certain about moving the movers to the top bar (yet).
I feel like the primary actions of the toolbar should be on the leftest side.

After playing around a bit...

I have an alternative interaction pattern idea that tried to lighten the UI re: movers.

Screen Capture on 2019-12-06 at 12-20-18

(I don't know if this was tried previously).

The idea came from playing around with [Notion](https://www.notion.so/) (they don't use arrows in their mover though).

If this isn't the right thread to post this, please let me know!

These are my initial thoughts :)

As always, thank you!

**Comment by youknowriad** [comment]
@ItsJonQ The idea of this PR is to try to move towards the direction of  #18667 where the block movers are in the toolbar but hidden under a hover effect.

Your proposal is a good alternative but it seems to move us into a different direction than the one suggested in  #18667 so maybe we should figure out the path forward there?

**Comment by ItsJonQ** [comment]
> so maybe we should figure out the path forward there?
Makes sense! I'll repost my comment there :)

**Comment by ellatrix** [comment]
@youknowriad What's needed to move this forward? Any way I can help?

**Comment by youknowriad** [comment]
I think we need a design direction here. What's acceptable to ship? Should we implement the "hover to show" effect first? Should we combine the three buttons into one (drag, up and down)?

**Comment by karmatosed** [comment]
Here are a couple of explorations. First, literally moving the simplified handles.



The issue around this is the benefit of a larger hit area is lost because of the thinner toolbar. Here is a comparison of potential height increase:



Notice here the starter border I also moved to the block, which could be worth exploring later.

cc @jasmussen for some ideas on this.

**Comment by jasmussen** [comment]
One of the challenges of putting both movers in the toolbar above and below each other, is that it makes the hit area very small, height wise. Currently the toolbar is 36px tall, which means you'd have 18px hit area for each half of the unified control. On mobile it's recommended touch targets aren't smaller than 24x24, and the controls in master right now are 24x28:

Screenshot 2019-12-10 at 10 20 51

In #18667 a 48px tall toolbar is being explored, which would give us 24px height, but even then it'd be nice with a solution that makes them even bigger, which is why that ticket also explores showing them when the block type is selected or hovered. 

An interim solution to moving them to the toolbar today _could_ be exploring vertically expanding the hit aa for the controls, invisibly. Fake it, so to speak, like this:

Screenshot 2019-12-10 at 10 26 49

The button invisibly extends beyond the boundaries. However faking it like this does not feel right, and another option is to do the "genie effect" where it increases size when you hover the control. I.e. this:

Screenshot 2019-12-10 at 10 29 16

**Comment by youknowriad** [comment]
As a first step,  in the last commit, I've made the mover buttons as the draggable element and removed the explicit one. Let me know if that's acceptable for you and whether I should move forward with trying to "merge" the buttons like suggested above.

**Comment by jasmussen** [comment]
GIF:

dragger

I think that's an excellent first step. I think once you actually press and hold, as opposed to just click, we can potentially change the visual appearance of the mover control to be that of a drag handle. 

The Mickey Mouse hand is plenty indicator, IMO.

**Comment by youknowriad** [comment]
> I think that's an excellent first step. I think once you actually press and hold, as opposed to just click, we can potentially change the visual appearance of the mover control to be that of a drag handle.

The problem here is that right now we only copy the "block content" on the draggable clone which means there's no toolbar there and this uses DOM cloning techniques which means moving the toolbar to a Popover like suggested in #18779 will make this impossible to do.

**Comment by youknowriad** [comment]
it might be possible if we change how drag & drop work and move the actual block instead. As you know, I already tried this in a PR and it's very hard to make it work. Not impossible though. So I wonder if we can find an alternative for now otherwise, this might block for some time.

**Comment by jasmussen** [comment]
Thanks for clarifying, was about to ask about the alternate method of drag and drop that did not clone. 

One quick point of clarification — on the draggable clone, could we re-create a faux drag toolbar where the toolbar used to be? One whose sole purpose is to show that changed drag state?

**Comment by youknowriad** [comment]
> . on the draggable clone, could we re-create a faux drag toolbar where the toolbar used to be? One whose sole purpose is to show that changed drag state?

Potentially, but with very very hacky code as the Drag & Drop is a generic component, this makes it very specific to blocks.

**Comment by youknowriad** [comment]
I'd rather try the new Draggable approach again though. Maybe this time, I'll have better luck.

**Comment by jasmussen** [comment]
I don't think the dragging state for the toolbar is necessary for this. I consider drag and drop to be a secondary affordance, and one that will have a larger hit area with this change. 

By the way, the select mode could also be a playground to make the entire block the draggable hit area ;)

**Comment by ZebulanStanphill** [comment]
I think making the up/down buttons double as a drag handle was a good decision. It both increases the size of the drag handle while reducing the total size of the movement controls.

I have noticed 2 issues in this branch right now.

First of all, the up/down buttons can't be used as a drag handle when they are disabled, meaning that a lone nested block can't be dragged.
image

Second, there are no move controls in the multi-block toolbar.
image

**Comment by ZebulanStanphill** [comment]
Side note: I think it's kind of funny how similar the block toolbar in this PR is compared to [this mockup from a year and a half ago](https://github.com/WordPress/gutenberg/issues/6224#issuecomment-405187124).

**Comment by youknowriad** [comment]
> Side note: I think it's kind of funny how similar this block toolbar is compared to this mockup from a year and a half ago.

Definitely, there's a lot of things we go back and forth on with exploration and feedback. Also as we move and add things, we enable new possibilities. For example these days the toolbars have collapsing dropdowns which gives us more room to try this.

**Comment by youknowriad** [comment]
Happy to continue here and fix those issues raised by @ZebulanStanphill but before investing further, I want to confirm that the current state is a direction we're ok landing? Should I push it further and combine in one button?

**Comment by youknowriad** [comment]
In the last commits, I fixed the issues raised by @ZebulanStanphill and I also tried showing the buttons together. What do you think?

**Comment by talldan** [comment]
Although it looks like I did ok in this gif, I think I found the button a bit too small to comfortably use.

mini-mover

**Comment by ellatrix** [comment]
What if the buttons expand vertically as you hover over them, as @jasmussen described in https://github.com/WordPress/gutenberg/pull/18685#issuecomment-563945427?

**Comment by pablohoneyhoney** [comment]
If we eventually increase the height of the toolbar, the vertical arrows work better, as they have an area of 24x48 (in case we use 12 as a base). Alternately the movers could expand on hover but might create other challenges in tight spaces and so on. 

1. If we are able to combine the ellipsis functions into the block type (as suggested by @ZebulanStanphill [here](https://github.com/WordPress/gutenberg/issues/18667#issuecomment-564708742) which we need to explore), the movers on the right work nicely. 
2. The UI conflict is with _More editing tools arrow_ (which could be replaced by an ellipsis if 1.)
3. We could allow dragging the toolbar also from the block type. 

Screen Shot 2019-12-12 at 10 10 21 AM

**Comment by ZebulanStanphill** [comment]
@ellatrix I tried out the branch just now. The tiny buttons are technically usable, but definitely not comfortable to use with a mouse. Keep in mind that some people with motor impairments are likely to struggle with controls that small. The expand-on-hover thing might help, but something else would have to be done for mobile.

Perhaps we should try a single button (that doubles as a drag handle) that simply opens up a sub-toolbar (either above, below, or replacing the main toolbar) containing the up and down controls. Actually, we could use this freedom of space to add additional controls such as move-to-top, move-to-bottom, and move-out-of-nested-context (something currently not possible without drag-and-drop).

Here's a text art example of what I mean:

_Click the ↕ button…_
[ ↕ ][ ¶ ][ ☰ ][ B I ∞ … ]

_…and it replaces the current toolbar (or opens another toolbar above/below) with something like:_
[ ↑ ↓ ↟ ↡ ]

**Comment by ItsJonQ** [comment]
@pablohoneyhoney
Hallo!! I just updated the prototypes with Pablo's suggestion for move/dragging.

https://tz3ir.csb.app/#/toolbar-base

Screen Capture on 2019-12-12 at 11-25-52

Pardon if the click -> drag interaction is bit wonky. The logic is getting complex in there, haha.

> Here's a text art example of what I mean:

@ZebulanStanphill I really appreciate the Text art! As someone who absorbs information visually, it helps a lot. I think clicking the arrows to reveal a movement toolbar is an interesting idea. Would be interesting to flush that out in Figma :)

**Comment by youknowriad** [comment]
I increased the toolbar size in the last commit along with some needed refactorings for the toolbar styles.
I think it works, you need to give some minutes to your muscle memory to adapt but once passed the shock, I think it feels good. What do you think?

**Comment by talldan** [comment]
@youknowriad That feels pretty good! It'd be cool to see the horizontal style movers in the nav block too.

The bigger toolbar feels quite nice too, though I wonder if spacing between blocks might need to increase to reduce the amount of overlap with the block above?

**Comment by youknowriad** [comment]
> spacing between blocks might need to increase to reduce the amount of overlap with the block above?

That's a theme issue as the spacing can be different from a theme to another. I think the UI should be able to adapt to any theme/spacing. There's definitely overlap but I thiink the solution here is more related to the "timing" of the toolbar (when it shows up and disappear). I think we do aa good job in that area but we can definitely do better. (maybe having an explicit way to close the toolbar)

**Comment by ZebulanStanphill** [comment]
Wow, those are some big buttons. And yet I think the arrow buttons are still too small. If we're increasing the button size to work better on mobile, then it's not doing much to improve the arrow buttons, which are still smaller than even the current button size in `master`. I still say they should be separate full-size buttons, even if they're hidden behind a sub-toolbar or something.

Also, if we're increasing the size of the toolbar buttons to 48px, would we also increase the size of buttons everywhere else?

**Comment by youknowriad** [comment]
@ZebulanStanphill worth noting the last commit doesn't affect the mover in mobile or the top toolbar, it's just for the contextual toolbar.

**Comment by talldan** [comment]
This PR also fixes #18195.

**Comment by jasmussen** [comment]
This is an excellent exploration. Movers in the toolbar benefit on many levels:

- A unified UI and a single toolbar to learn.
- Consistent placement regardless of block.
- Consistent interface regardless of direction of movement, no need for separate controls as in the navigation block.

Despite appearances, this also makes the mover control bigger than what's shipping in the software

Screenshot 2019-12-16 at 11 25 54

Screenshot 2019-12-16 at 11 25 03

Mobile will benefit from the unified toolbar as well, and we'll adjust the size to be appropriate for touch.

I recorded a video walking through this: https://cloudup.com/cL7gmj8Tpaq

A few take-aways from that:

- The margins and paddings need massaging. Dropdowns should have the same size, up/down arrows should be optically centered.
- Multiple sequential buttons in a group (such as bold, italic, link) needs a heuristic to look optically balanced. [Buttons other than the first and last have a different padding making their effective hit area 36x48px](https://user-images.githubusercontent.com/1204802/70900323-ddb1f800-1ff8-11ea-9a80-3863fbc99899.png).
- Toolbar in wide and floats need some adjustment.
- Quickly clicking the mover control blocks sometimes "jump" and you lose place.
- When moving a paragraph beyond an image makes the toolbar be obscured (lower z-index) until the block stops moving.
- Not sure how to reproduce, but sometimes moving upwards the block is moved incorrectly to the end, and you lose your place.

Items 1-3 I can help fix. But before I do it's worth figuring out how to approach this. Notably the toolbar is not yet detached, so it's still cropped in narrow contexts, which is more pronounced with the bigger toolbar:

Screenshot 2019-12-16 at 11 45 42

While as a prototype this PR feels like a big step forwards in terms of code and interface reduction, it seems in order to ship this we need to first figure out the mover issues. 

We also need to figure out whether it's possible to ship a version of movers in the toolbar that don't need the larger footprint. Alternatively it seems like we need to wait for the detached toolbar (#18779) to land, so as to not make the toolbar cropping issue worse.

**Comment by ZebulanStanphill** [comment]
@jasmussen
>Despite appearances, this also makes the mover control bigger than what's shipping in the software

This surprised me. I guess the mover control is better in this PR than in `master` after all. I'm still not convinced that increasing the overall toolbar size is a good way to accomplish this, though. (Especially considering that the inline formatting tools aren't even 48px wide yet.) I haven't done the math, but I wonder if just adding two separate mover buttons might take up less space overall than increasing the whole toolbar size to accomodate stacked mover buttons.
 
>Quickly clicking the mover control blocks sometimes "jump" and you lose place.

This is an issue in `master` as well, so I don't think it should be a blocker for this PR.

I do agree that it would probably be a good idea to detach the toolbar from the left block border before making it larger.

**Comment by shaunandrews** [comment]
I thought I would hate the taller toolbar, but I don't mind it. I do like the movers in the toolbar, and I like the vertical orientation — so much so that I'd like to see that same layout for the movers when in "Top Toolbar" mode. Right now it looks like this:

image

**Comment by shaunandrews** [comment]
Related, I prefer the way the icon hover works when in "Top Toolbar" mode:

image

The clickable area with the dark, rounded border doesn't fill the vertical space. However, in this PR, the taller toolbar also increase the height of the rounded border on hover:

image

Visually, I prefer the space around the hover area — it helps set the icon buttons apart from their container and avoids the awkward double-border affect we see in `master`.

**Comment by youknowriad** [comment]
@shaunandrews All good points, that said changing the TopToolbar and the hover style of the contextual toolbar will probably take a lot of CSS and changes and that PR is huge already.

If possible I'd prefer to wait for the blocker in #18779 and land it as is with follow-up PRs.

**Comment by youknowriad** [comment]
Rebased that PR now that #18779 is merged. The truncated toolbar is not fixed yet. If I'm not wrong @ellatrix had plans for it.

**Comment by youknowriad** [comment]
We merged #19322 (toolbar position) into this PR, so the last blocker is gone. I think we're ready to ship this PR.

File: packages/block-editor/src/components/block-switcher/style.scss [hunk 1]
```diff
@@ -8,13 +8,12 @@
...
 	height: $icon-button-size;
-	padding: 3px;
 }
 
 .components-button.block-editor-block-switcher__no-switcher-icon {
-	width: $icon-button-size + 6px + 6px;
+	width: $icon-button-size;
 
-	.block-editor-block-icon {
+	.block-editor-blocsk-icon {
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-switcher/style.scss]: Looks like a typo here.

**Comment by ellatrix** [comment]
The toolbar look huge in my opinion, but ok if design approves.

This button is smaller than the others and it looks like it's positioned closer to the edge than the bold button.

Screenshot 2020-01-14 at 10 57 02

**Comment by jasmussen** [comment]
> The toolbar look huge in my opinion, but ok if design approves.

I'm working currently on #19344 with the same size toolbar, and with the right overall treatment, it feels great. Feel free to check out the branch, but here's a GIF:

**Comment by youknowriad** [comment]
Yes, design-wise this PR is a step towards #19344

**Review by ellatrix — APPROVED** [review]
Looks good to me, but would appreciate another check from design.

**Comment by ellatrix** [comment]
Another small issue I see: when I drag a block, one of the block mover tooltips pops up and stays on he screen while I'm dragging blocks.

**Comment by jasmussen** [comment]
Yep. I wonder if it's too soon to do merge this. The exact behavior of the mover control — permanently visible or visible in certain focus situations — is not yet fleshed out. And without the new look, it DOES look big and a lot of the focus styles are off.

Should 19344 be a blocker for this one?

**Comment by youknowriad** [comment]
19344 is built on top of this branch, so it's already included there, we should just rebase it over this branch (I'll do it).

**Comment by jasmussen** [comment]
Let me know if/when you do that and I'll postpone pushing until then.

**Comment by youknowriad** [comment]
@jasmussen sorry, just saw the message, I already rebased

**Comment by youknowriad** [comment]
This was merged into #19344 and will land with it. So let's close this one.

**Pull request closed without merging**